src/common/hive_partitioning.cpp

```diff
@@ -0,0 +1,107 @@
+#include "duckdb/common/hive_partitioning.hpp"
+#include "duckdb/optimizer/statistics_propagator.hpp"
+#include "duckdb/planner/table_filter.hpp"
+#include "duckdb/execution/expression_executor.hpp"
+#include "duckdb/optimizer/filter_combiner.hpp"
+#include "duckdb/planner/expression_iterator.hpp"
+#include "re2/re2.h"
+
+#include <iostream>
+
+namespace duckdb {
+
+static std::map<string, string> GetKnownColumnValues(string &filename, bool filename_col, bool hive_partition_cols) {
+	std::map<string, string> result;
+
+	if (filename_col) {
+		result["filename"] = filename;
+	}
+
+	if (hive_partition_cols) {
+		auto partitions = HivePartitioning::Parse(filename);
+		for (auto &partition : partitions) {
+			result[partition.first] = partition.second;
+		}
+	}
+
+	return result;
+}
+
+// Takes an expression and converts a list of known column_refs to constants
+static void ConvertKnownColRefToConstants(unique_ptr<Expression> &expr, map<string, string> &known_column_values) {
+	if (expr->type == ExpressionType::BOUND_COLUMN_REF) {
+		auto lookup = known_column_values.find(expr->alias);
```

Comment:
Looking up based on alias is a bit fragile, can't we use the column index into the set of columns provided by the table scan here?

Reply:
good point!